### PR TITLE
Added new PyramidImage class. Added optimized derivatives in RasterMonochrome.

### PR DESCRIPTION
**Pull request by niklasborwell**

### CODE REVIEW

**Comment by niklasborwell** [comment]
This is gonna be a bit awkward because I'm including stuff from Bettina and Ziquan in both pull requests.

**Comment by niklasborwell** [comment]
Found a lot of leakage in the new pyramid image class. I'm working on it.

File: source/draw/gpu/PyramidImage.ooc [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+		result: RasterMonochrome
+		size := this _scaledResolution
+		offset := 0
+		for (i in 0..level) {
+			offset += size area
+			size /= 2
+		}
+		match(level) {
+			case 0 =>
+			 result = this base
```
Comment (davidhesselbom) [inline source/draw/gpu/PyramidImage.ooc]: Indentation

File: source/draw/gpu/PyramidImage.ooc [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+					this _level3 = this _getRepresentationSlice(offset, size)
+					this _level3 referenceCount increase()
+				}
+				result = this _level3
+			case 4 =>
+				if (this _level4 == null) {
+					this _level4 = this _getRepresentationSlice(offset, size)
+					this _level4 referenceCount increase()
+				}
+				result = this _level4
```
Comment (davidhesselbom) [inline source/draw/gpu/PyramidImage.ooc]: There's no default case.

File: source/draw/gpu/PyramidImage.ooc [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+	_scaledResolution: IntSize2D
+	scaledResolution ::= this _scaledResolution
+	_representationRows: Int
+	representationRows ::= this _representationRows
+
+	init: func ~fromRepresentation (representation: GpuMonochrome, =_pyramidLevels, =_scaledResolution, =_representationRows) {
+		super(this _scaledResolution)
+		this _validRaster = false
+		if (this _fullRepresentationGpu != null) this _fullRepresentationGpu recycle()
+		this _fullRepresentationGpu = representation
```
Comment (davidhesselbom) [inline source/draw/gpu/PyramidImage.ooc]: indentation

File: source/draw/gpu/PyramidImage.ooc [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+	}
+	_getRepresentationSlice: func(offset: Int, size: IntSize2D) -> RasterMonochrome {
+		result := RasterMonochrome new(this fullRepresentation buffer slice(offset, size area), size)
+		result
```
Comment (davidhesselbom) [inline source/draw/gpu/PyramidImage.ooc]: Just `RasterMonochrome new(this fullRepresentation buffer slice(offset, size area), size)` would do nicely, there's no reason to use a `result` variable for one-liners.

File: source/draw/gpu/PyramidImage.ooc [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+	_toRaster: func{
+		if (this _fullRepresentation != null)
+			this _fullRepresentation referenceCount decrease()
+		this _fullRepresentation = fullRepresentationGpu toRaster() as RasterMonochrome
+		this _fullRepresentation referenceCount increase()
+		this _validRaster = true
+	}
+	__destroy__: func {
+		super()
+	}
```
Comment (davidhesselbom) [inline source/draw/gpu/PyramidImage.ooc]: Is this function necessary?

File: source/draw/gpu/PyramidImage.ooc (hunk 2 above)
Comment (niklasborwell) [inline source/draw/gpu/PyramidImage.ooc]: Just for discussions sake, we don't normally do any exception handling or checking of input values. So why should we create default cases?

File: source/draw/gpu/PyramidImage.ooc (hunk 5 above)
Comment (niklasborwell) [inline source/draw/gpu/PyramidImage.ooc]: Yes. It calls super().

File: source/draw/gpu/PyramidImage.ooc (hunk 2 above)
Comment (niklasborwell) [inline source/draw/gpu/PyramidImage.ooc]: Fixed.

File: source/draw/gpu/PyramidImage.ooc (hunk 5 above)
Comment (davidhesselbom) [inline source/draw/gpu/PyramidImage.ooc]: Ah. Carry on.

**Comment by niklasborwell** [comment]
Now with less PyramidImage.

**Pull request merged**